feat: :sparkles: added docker support

jqshuv opened the pull request:
I've added Docker support.

Things I add:
* `Dockerfile` to build the bot.
* `docker-compose.yml` to deploy it easy without an external mongodb database.
* `enviroment.d.ts` to fix `process.env["ENV"]` typings

jqshuv committed:
feat: :sparkles: added docker support

vermium commented on src/types/enviroment.d.ts:
```diff
@@ -0,0 +1,27 @@
+      DB_USER: string;
+      DB_PASSWORD: string;
+      DISCORD_TOKEN: string;
+      DISCORD_CLIENT_ID: string;
```
You could use Discord Types here. By importing with ``import { Snowflake } from "discord.js";`` you could you Snowflake as a type, this way we can preserve *.ts config file types!

jqshuv replied:
Ok, im looking at this!

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    restart: unless-stopped
+    environment:
+      - MONGO_INITDB_ROOT_USERNAME=mongouser
+      - MONGO_INITDB_ROOT_PASSWORD=mongopasword
```
I¨m kinda sure you need to add this to ``mongodb:``. ``command: --auth`` to enable the authentication system, I might be wrong.

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+      - HOSTER_NAME=YOUR_HOSTER_NAME
+      - NODE_ENV=development
+    stdin_open: true
+    tty: true
```
Could you maybe also make the ability to pass troguh logs from root directory/logs to like /logs as a docker volume?

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+FROM node:16
+
+LABEL maintainer="me@jqshuv.xyz"
```
Could you change ``maintainer`` to ``xyter@zyner.org``?

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+WORKDIR /app
+
+LABEL maintainer="me@jqshuv.xyz"
```
Could you change ``maintainer`` to ``xyter@zyner.org``?

vermium commented on src/types/enviroment.d.ts:
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2022 Joshua Schmitt
```
Could you remove your comment here? Because this source code is under GNU General Public License v3.0 as shown in https://github.com/ZynerOrg/xyter/blob/main/LICENSE. Also by signing your commits (git with gpg keys)  you basicly sign your work.

jqshuv replied:
Oh sorry it's my vscode addon I forgot to turn it off. As I said it's new to work with other people 😅

vermium replied:
I understand, no problem! Could you link to that extension? I may want to use it for some other projects :)

jqshuv replied:
[Here](https://marketplace.visualstudio.com/items?itemName=ymotongpoo.licenser) we go.

vermium replied:
Thank you 👍

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    build: .
+    restart: unless-stopped
+    environment:
+      # You can leave this as-is.
```
Could you change from DB to MONGO? Also add the ability to use MONGO_URL instead of seperating them?

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    stdin_open: true
+    tty: true
+  mongodb:
+    image: mongo:latest
```
Could you pin mongo to the latest pinned version? [This one](https://hub.docker.com/layers/mongo/library/mongo/4.2.21/images/sha256-68f432c24a2086e234f8cfaa531963c7fcba946f96d60af2f44ac99477f0f780?context=explore) is the latest one right now.

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+  app:
+    depends_on:
+      - mongodb
+    build: .
```
Could you change this from ``build: .`` to ``image: zyner/xyter:2022.07.0`` for next release.

jqshuv replied:
Yes, but you shouldn't forget to push the image before releasing.

vermium replied:
True! Make it ``image: zyner/xyter:latest`` instead. This way I can automate Docker Image Build via a Github Action. 👍

jqshuv replied:
You can look at the build_image step [here](https://github.com/Unately/uBingo/blob/master/.github/workflows/release.yml) I've done something like that already myself.

vermium replied:
Would be super cool if you could make a PR with that too 😅 To push dev versions and production versions, like tags as ``2022.07.0`` or ``dev-2022.07.0`` you get the point, with images as the branch name.

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+      - DISCORD_CLIENT_ID=YOUR_CLIENT_ID
+      - DISCORD_DEV_GUILD_ID=YOUR_DEV_GUILD_ID
+      - HOSTER_NAME=YOUR_HOSTER_NAME
+      - NODE_ENV=development
```
Could you make this one to production by default?

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    ports:
+      - $MONGODB_LOCAL_PORT:27017
+    volumes:
+      - db:/data/db
```
Could you make this as a ``bind mount``? I recommend keeping it outside of the bot code. So like ``root directory/mongodb`` or ```root directory/database`` or something in that way. Just so that the docker image isn't accessing it (app container).

jqshuv replied:
The volume don't access the app directory 😊 the ```db``` is a volume which is declared at the top it is saved at the local machine, the path ist something like ```/var/lib/docker/volumes```. And it saves the folder ```/data/db``` in that volume.

vermium replied:
I know but, I think many users will accidently remove their database if they updating Docker in a wrong way. I think it's better to do a bind mount, since they only need the database as a volume, this could be mounted for example to ``./database:/data/db`` or something by default

vermium commented on package.json:
```diff
@@ -21,6 +21,9 @@
   },
   "author": "Vermium Sifell <vermium@zyner.org> (https://zyner.org)",
+  "contributors": [
+    "Joshua Schmitt <me@jqshuv.xyz> (https://jqshuv.xyz)"
```
This is awesome, I forgot to use contributors field!

vermium commented on src/config_docker/encryption.ts:
```diff
@@ -0,0 +1,6 @@
+
+// Encryption secret (strictly 32 length)
+export const secretKey =
+  process.env["SECRET_KEY"] || "h/f#Ts8w5sch5L:J*_gPW)]$'!4K.K.-";
```
Please do not include a default secret key, use something longer than 32 to throw an error to bot hoster. To keep the data safer.

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+WORKDIR /app/build
+
+CMD [ "npx", "nodemon" ]
```
Since this is inside of a container, I think a better way to do it is by using ``node .`` and exit the container if bot is crashing. To make use of the ``restart: unless-stopped`` docker feature, for example.

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+RUN npm -v > /app/npm.log
+RUN node -v > /app/node.log
+
+WORKDIR /app/build
```
Would look better if you make build directory go straight to /app instead. I think this is possible if you don't set ``WORKDIR /app`` at the top.

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+RUN npx tsc
+
+RUN npx tsc -v > /app/tsc.log
```
This should be inside of ``logs/``

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+RUN npx tsc
+
+RUN npx tsc -v > /app/tsc.log
+RUN npm -v > /app/npm.log
```
This should be inside of ``logs/``

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+RUN npx tsc -v > /app/tsc.log
+RUN npm -v > /app/npm.log
+RUN node -v > /app/node.log
```
This should be inside of ``logs/``

jqshuv replied:
Yes, first it was only to send you the versions, due to the error. But I can do this of course.

vermium replied:
Maybe even better to post it into ``logs/package.versions`` or something. This way it's easy to help users that has problems. But remove it if you want to, since I wont publish a non working image 😊

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+COPY package* .
+RUN npm install
+
+COPY . .
```
Is this really necessary? You are copying from current dir to current dir., right?

jqshuv replied:
No, I copy from the current local directory to the docker workdir.

vermium replied:
Oh I see!

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,30 @@
+WORKDIR /app
+
+LABEL maintainer="me@jqshuv.xyz"
+LABEL org.opencontainers.image.source https://github.com/ZynerOrg/xyter
```
Why is this ``org.opencontainers.image`` here? Is this something standard used in Dockerfiles?

jqshuv replied:
It's if you're pushing an image to the ghcr.io container registry, but as I read you're pushing to docker hub so I could remove it.

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+      - MONGO_INITDB_ROOT_USERNAME=mongouser
+      - MONGO_INITDB_ROOT_PASSWORD=mongopasword
+    ports:
+      - $MONGODB_LOCAL_PORT:27017
```
This port is not necessary. Since you are having both containers in same ``subnet`` they can access each other anyway. This will only be a security flaw if used by default.

jqshuv replied:
Yep, I forgot to remove that.

vermium commented on src/types/enviroment.d.ts:
```diff
@@ -0,0 +1,27 @@
+  }
+}
+
+export {};
```
Is exporting nothing required here? Asking because I don't know.

jqshuv replied:
Yes, read it on stackoverflow. And it actually don't export nothing, it export the file itself I guess.

vermium replied:
I think you can do, ``export declare global {`` I might be wrong

jqshuv replied:
This could be, I'm not into typescript typing that much

vermium replied:
Kinda new to declaration too

vermium reviewed (changes requested):
Thanks for your first contribution, I request some changes to the following, code. Thanks man!

vermium committed:
💥 migrate from *.ts to .env configs

vermium committed:
🔧 example env

jqshuv committed:
Merge pull request #1 from VermiumSifell/dev

Migration from *.ts to .env

jqshuv committed:
Merge pull request #2 from jqshuv/dev

Dev

vermium reviewed (approved):
LGTM

jqshuv committed:
fix: :rocket: fix and rewrote docker stuff

I reworked the docker files and added docker support for the new config system.

The old config system is no longer working

jqshuv committed:
style: :lipstick: fix style

vermium commented on .gitignore:
```diff
@@ -140,3 +140,8 @@ dist
+
+
+# Docker
+database
```
Looks great as a extra security to not pass through database while building image! 👍

vermium commented on .gitignore:
```diff
@@ -140,3 +140,8 @@ dist
+
+# Docker
+database
+docker-compose.local.yml
```
Any idea how to change so that contributors are not accidently pushing their credentials?

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM node:16
+
+LABEL maintainer="xyter@zyner.org"
```
Thanks! 👍

vermium commented on Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+WORKDIR /app/build
+
+CMD [ "node", "." ]
```
Will the bot crash the docker container if bot crashes?

vermium commented on docker-compose.dev.yml:
```diff
@@ -0,0 +1,33 @@
+      - DISCORD_TOKEN=YOUR TOKEN
+      - DISCORD_CLIENT_ID=YOUR CLIENT ID
+      - DISCORD_GUILD_ID=YOUR GUILD ID
+      ### LEAVE THIS AS IT IS! ###
```
I think it's better to not say anything, since users **SHOULD** change this, as it's insecure to not do it. Maybe adding a .env for the docker compose or something. So you can use variables and make users configure environments in a file. For example using these files it's possible. ``docker-compose.yml`` ``.env``. I don't know.

vermium commented on docker-compose.dev.yml:
```diff
@@ -0,0 +1,33 @@
+      - REPUTATION_TIMEOUT=86400
+      - BOT_HOSTER_NAME=Zyner
+      - BOT_HOSTER_URL=https://xyter.zyner.org/customization/change-hoster
+      - NODE_ENV=production
```
Since this is the development file, it should be ``development``

vermium commented on docker-compose.dev.yml:
```diff
@@ -0,0 +1,33 @@
+    image: mongo:4.2.21
+    restart: unless-stopped
+    volumes:
+      - ./logs:/app/build/logs
```
This should be on ``app`` container ant not ``mongodb``?

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+      - DISCORD_TOKEN=YOUR TOKEN
+      - DISCORD_CLIENT_ID=YOUR CLIENT ID
+      - DISCORD_GUILD_ID=YOUR GUILD ID
+      ### LEAVE THIS AS IT IS! ###
```
I think it's better to not say anything, since users **SHOULD** change this, as it's insecure to not do it. Maybe adding a .env for the docker compose or something. So you can use variables and make users configure environments in a file. For example using these files it's possible. ``docker-compose.yml`` ``.env``. I don't know.

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    image: mongo:4.2.21
+    restart: unless-stopped
+    volumes:
+      - ./logs:/app/build/logs
```
This should be on ``app`` container ant not ``mongodb``?

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+  app:
+    depends_on:
+      - mongodb
+    image: zyner/xyter:2022.07.0
```
I think better is to say ``zyner/xyter:latest``

vermium commented on docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+    stdin_open: true
+    tty: true
+  mongodb:
+    image: mongo:4.2.21
```
I think I regret this change, I think it's better to say ``mongo:latest``. To keep security up to date for users. And make them change it themself if they want to.

vermium reviewed (changes requested):
Well done, a little more changes please 😊 Soon it's perfect

jqshuv committed:
fix: :construction: fix review files

jqshuv committed:
fix: :construction: done review

jqshuv committed:
fix: :art: set docker-compose version

jqshuv committed:
fix: :bug: fix env to docker-compose file

vermium committed:
Merge branch 'dev' into main

vermium reviewed (approved):
LGTM
